unit-test/broker-protection.js

```diff
@@ -0,0 +1,422 @@
+                // eslint-disable-next-line no-template-curly-in-string
+                url: 'https://example.com/profile/search?fname=${firstName}&lname=${lastName}&state=${state}&city=${city|snakecase|downcase}&fage=${age}'
+            }, userData2)
+
+            expect(result).toEqual({ url: 'https://example.com/profile/search?fname=John&lname=Smith&state=ny&city=west_montego&fage=24' })
+        })
+
+        it('should build hyphenated url when given hyphenated state', () => {
+            const result = replaceTemplatedUrl({
+                id: 0,
+                // eslint-disable-next-line no-template-curly-in-string
+                url: 'https://example.com/profile/search?fname=${firstName}&lname=${lastName}&state=${state|stateFull|hyphenated}&city=${city}&fage=${age}'
+            }, userData2)
+
+            expect(result).toEqual({ url: 'https://example.com/profile/search?fname=John&lname=Smith&state=New-York&city=West+Montego&fage=24' })
+        })
+
+        it('should build url when given valid data and age range', () => {
+            const result = replaceTemplatedUrl({
+                id: 0,
+                // eslint-disable-next-line no-template-curly-in-string
+                url: 'https://example.com/profile/search?fname=${firstName}&lname=${lastName}&state=${state}&city=${city}&fage=${age|ageRange}',
+                ageRange: ['18-30', '31-40', '41-50']
+            }, userData)
+
+            expect(result).toEqual({ url: 'https://example.com/profile/search?fname=John&lname=Smith&state=il&city=Chicago&fage=18-30' })
+        })
+
+        it('should error when given an invalid action', () => {
+            const result = replaceTemplatedUrl({
+                id: 0,
+                url: null
+            }, userData)
+
+            expect(result).toEqual({ error: 'Error: No url provided.' })
+        })
+
+        it('should accept any inputs and not crash', () => {
+            fc.assert(
+                fc.property(
+                    fc.oneof(
+                        fc.anything(),
+                        fc.record({
+                            url: fc.anything()
+                        })
+                    ),
+                    fc.oneof(
+                        fc.anything(),
+                        fc.dictionary(fc.string(), fc.oneof(fc.string(), fc.integer(), fc.boolean()))
+                    ),
+                    (action, userData) => {
+                        const result = replaceTemplatedUrl(action, userData)
+                        expect('url' in result || 'error' in result)
+                        if ('error' in result) {
+                            expect(typeof result.error).toEqual('string')
+                        }
+                        if ('url' in result) {
+                            const url = new URL(result.url)
+                            expect(url).toBeDefined()
+                        }
+                    }
+                )
+            )
+        })
```

Comment:
these property-based tests were added to help me stress-test the URL replacement stuff (interpolating data). It helped me catch a couple of bugs, and in general it would be nice to have fast-check around in C-S-S to use for other fuzzing-style tests

Reply:
Are we using fast-check anywhere else in JS land? If not I think it's an open a new tech thing as it's sizable dependency.

Reply:
AFAIK it's the most widely used in the JS eco-system (likely not a huge space, given that many don't reach for this kind of testing)

We have it in autofill (password generation) - but I believe that was agreed during the code-review, rather than a new-tech discussion.

We could remove these tests (I applied the fixes they raised) and then come back after we've done the official channels :)

Reply:
> AFAIK it's the most widely used in the JS eco-system 

I meant our ecosystem.

> We could remove these tests (I applied the fixes they raised) and then come back after we've done the official channels :)

I think we're good if used elsewhere and it's only for testing.